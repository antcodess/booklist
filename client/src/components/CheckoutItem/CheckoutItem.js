import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import { addItem, removeItem, clearItem } from '../../redux/cart/cart.actions';

import './CheckoutItem.scss';

const CheckoutItem = ({ item, addItem, removeItem, clearItem, history }) => {
  const { imageUrl, title, quantity, price, routeName } = item;
  return (
    <div className='checkout-item'>
      <div className='image-container' onClick={() => history.push(`/${routeName}`)}>
        <img src={imageUrl} alt='book' />
      </div>
      <span className='title' onClick={() => history.push(`/${routeName}`)}>{title}</span>
      <span className='quantity'>
        <div className='arrow' onClick={() => removeItem(item)}>&#10094;</div>
        <span className='value'>{quantity}</span>
        <div className='arrow' onClick={() => addItem(item)}>&#10095;</div>
      </span>
      <span className='price'>${price}</span>
      <div className='remove' onClick={() => clearItem(item)}>&#10005;</div>
    </div>
  );
};

const mapDispatchToProps = dispatch => ({
  addItem: item => dispatch(addItem(item)),
  removeItem: item => dispatch(removeItem(item)),
  clearItem: item => dispatch(clearItem(item))
});

export default connect(null, mapDispatchToProps)(withRouter(CheckoutItem));