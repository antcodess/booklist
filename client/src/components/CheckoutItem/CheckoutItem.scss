.checkout-item {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr 1fr;
  grid-gap: 10px;
  margin-top: 40px;

  .image-container {
    cursor: pointer;

    img {
      width: 100%;
      height: auto;
    }
  }
  .title,
  .quantity,
  .price,
  .remove {
    margin: auto 0;
  }

  .title {
    cursor: pointer;
  }

  .quantity {
    display: flex;
    justify-content: center;

    .arrow {
      cursor: pointer;
    }

    .value {
      margin: 0 10px;
    }
  }

  .remove {
    cursor: pointer;
  }
}