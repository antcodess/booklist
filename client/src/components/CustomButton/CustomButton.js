import React from 'react';

import './CustomButton.scss';

const CustomButton = ({ children, isGoogleSignIn, isRegularSignIn, ...otherProps }) => {
  return (
    <button 
    className={`custom-button ${isGoogleSignIn ? 'google-sign-in' : ''}${isRegularSignIn ? 'regular-sign-in' : ''}`} 
    {...otherProps}>{ children }</button>
  );
};

export default CustomButton;