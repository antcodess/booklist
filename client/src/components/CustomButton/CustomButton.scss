.regular-sign-in {
  margin-bottom: 5px;
  &:hover {
    background-color: black;
    opacity: 0.9;
  }
}

.google-sign-in {
  background-color: #0080DE;

  &:hover {
    opacity: 0.9;
  }
}

@media only screen and (min-width: 950px) {
  .regular-sign-in {
    width: 40%;
  }
}

@media only screen and (min-width: 1100px) {
  .regular-sign-in, .google-sign-in {
    width: 49%;
  }
}