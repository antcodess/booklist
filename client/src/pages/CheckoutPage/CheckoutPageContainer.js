import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';

import { selectIsOrderLoading } from '../../redux/orders/orders.selectors';

import WithSpinner from '../../components/WithSpinner/WithSpinner';
import CheckoutPage from './CheckoutPage';

const mapStateToProps = createStructuredSelector({
  isLoading: selectIsOrderLoading
});

const CheckoutPageContainer = compose(
  connect(mapStateToProps),
  WithSpinner
)(CheckoutPage);

export default CheckoutPageContainer;
