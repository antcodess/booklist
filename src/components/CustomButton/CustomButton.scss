.regular-sign-in {
  margin-bottom: 5px;
}

.google-sign-in {
  background-color: #0080DE;

  &:hover {
    background-color: black;
  }
}