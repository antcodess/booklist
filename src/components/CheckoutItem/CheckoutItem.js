import React from 'react';
import { connect } from 'react-redux';
import { addItem, removeItem, clearItem } from '../../redux/cart/cart.actions';
import './CheckoutItem.scss';

const CheckoutItem = ({ item, addItem, removeItem, clearItem }) => {
  const { imageUrl, title, quantity, price } = item;
  return (
    <div className='checkout-item'>
      <div className='image-container'>
        <img src={imageUrl} alt='book' />
      </div>
      <span className='title'>{title}</span>
      <span className='quantity'>
        <div className='arrow' onClick={() => removeItem(item)}>&#10094;</div>
        <span className='value'>{quantity}</span>
        <div className='arrow' onClick={() => addItem(item)}>&#10095;</div>
      </span>
      <span className='price'>${price}</span>
      <div className='remove' onClick={() => clearItem(item)}>&#10005;</div>
    </div>
  );
};

const mapDispatchToProps = dispatch => ({
  addItem: item => dispatch(addItem(item)),
  removeItem: item => dispatch(removeItem(item)),
  clearItem: item => dispatch(clearItem(item))
});

export default connect(null, mapDispatchToProps)(CheckoutItem);