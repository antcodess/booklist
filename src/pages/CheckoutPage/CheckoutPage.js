import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { selectCartItems, selectCheckoutTotal } from '../../redux/cart/cart.selectors';

import BreadCrumbNav from '../../components/BreadCrumbNav/BreadCrumbNav';
import CheckoutItem from '../../components/CheckoutItem/CheckoutItem';
import StripeCheckoutButton from '../../components/StripeButton/StripeButton';

import './CheckoutPage.scss';

const CheckoutPage = ({ cartItems, total }) => {
  return (
    <React.Fragment>
      <BreadCrumbNav title='Checkout'/>
      <section className='checkout-page-container'>
        <h2>Checkout</h2>
        <div className='checkout-page'>
          <div className='checkout-header'>
            <div className='header-block'>
              <span>Product</span>
            </div>
            <div className='header-block'>
              <span>Title</span>
            </div>
            <div className='header-block'>
              <span>Quantity</span>
            </div>
            <div className='header-block'>
              <span>Price</span>
            </div>
            <div className='header-block'>
              <span>Remove</span>
            </div>
          </div>
          { cartItems.length ? 
              cartItems.map(item => <CheckoutItem key={item.id} item={item} />)
              : (<div className='checkout-message'><span>Your cart is empty.</span></div>)
          }
          <div className='total'>
            <span className='total-bold'>Total: <span>${total}</span></span>
          </div>
          <div className='test-warning'>
            <span>This app was created for testing/proof of concept purposes.</span>
            <span>*Please ONLY use the following test credit card for payments*</span>
            <span>4242 4242 4242 4242 - Exp: 01/20 - CVV: 123</span>
          </div>
          <div className='checkout-button'>
            <StripeCheckoutButton price={total} />
          </div>
        </div>
      </section>
    </React.Fragment>
  )
};

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
  total: selectCheckoutTotal
});

export default connect(mapStateToProps)(CheckoutPage);