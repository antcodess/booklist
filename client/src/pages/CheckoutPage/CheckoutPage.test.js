import React from 'react';
import { shallow } from 'enzyme';
import { CheckoutPage } from './CheckoutPage';
import CheckoutItem from '../../components/CheckoutItem/CheckoutItem';
import CustomButton from '../../components/CustomButton/CustomButton';
import StripeCheckoutButton from '../../components/StripeButton/StripeButton';
import { mapDispatchToProps } from './CheckoutPage';

describe('CheckoutPage Component', () => {
  let wrapper;
  let wrapper2;
  let mockCartItems;
  let mockTotal;
  let mockCurrentUser;
  let mockSetAlert;

  const mockAlert = {
    status: 'failure',
    message: 'Your cart is empty. Please add an item to complete checkout process.'
  }

  const mockAlertPayload = {
    type: 'SET_ALERT',
    payload: mockAlert
  }
  
  beforeEach(() => {

    mockCartItems = [{ id: 1}, { id: 2 }, { id: 3 }];
    mockTotal = 90;
    mockCurrentUser = { id: 1, name: 'Sample User'};
    mockSetAlert = jest.fn();
    

    const mockProps = {
      cartItems: mockCartItems, 
      total: mockTotal, 
      currentUser: mockCurrentUser, 
      setAlert: mockSetAlert
    };

    const mockProps2 = {
      cartItems: [], 
      total: mockTotal, 
      currentUser: mockCurrentUser, 
      setAlert: mockSetAlert
    };

    wrapper = shallow(<CheckoutPage { ...mockProps } />);
    wrapper2 = shallow(<CheckoutPage { ...mockProps2 } />);
  });

  it('should render CheckoutPage Component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render CheckoutItem components if there are cartItems', () => {
    expect(wrapper.exists(CheckoutItem)).toBe(true);
  });

  it('should not render CheckoutItem components if there are no cartItems', () => {
    expect(wrapper2.exists(CheckoutItem)).toBe(false);
  });

  it('should not render checkout empty message if there are cartItems', () => {
    expect(wrapper.exists('.checkout-message')).toBe(false);
  });

  it('should render checkout empty message if there are no cartItems', () => {
    expect(wrapper2.exists('.checkout-message')).toBe(true);
  });

  it('should render StripeCheckoutButton if there are cartItems', () => {
    expect(wrapper.exists(StripeCheckoutButton)).toBe(true);
  });

  it('should not render StripeCheckoutButton if there are no cartItems', () => {
    expect(wrapper2.exists(StripeCheckoutButton)).toBe(false);
  });

  it('should not render CustomButton if there are cartItems', () => {
    expect(wrapper.exists(CustomButton)).toBe(false);
  });

  it('should render CustomButton if there are no cartItems', () => {
    expect(wrapper2.exists(CustomButton)).toBe(true);
  });

  it('should call setAlert when Custom Button component is clicked', () => {
    wrapper2.find(CustomButton).simulate('click');
    expect(mockSetAlert).toHaveBeenCalled();
  });

  it('should dispatch setAlert action when clicked', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).setAlert(mockAlert);
    expect(dispatch.mock.calls[0][0]).toEqual(mockAlertPayload);
  });
});