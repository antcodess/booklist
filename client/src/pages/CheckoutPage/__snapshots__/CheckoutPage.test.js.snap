// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutPage Component should render CheckoutPage Component 1`] = `ShallowWrapper {}`;
